```{r setup & libraries}
# load and install corresponding librarys
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager", INSTALL_opts = '--no-lock')
if (!requireNamespace("MEGENA", quietly = TRUE))
  install.packages("MEGENA")

BiocManager::install("EnrichmentBrowser", ask = FALSE)
library("EnrichmentBrowser")
if (!isAvailable("KEGGREST"))
  BiocManager::install("KEGGREST", ask = FALSE, type = "source", checkBuilt = TRUE)

library("KEGGREST")
```

```{r parse sand save finding}
# library("EnrichmentBrowser")

db <- "kegg"
org <- "hsa"
kegg_db <- "pathway" # KEGG database anme ie. pathway, brite, module, orthology, genome, genes
# list can be found at (https://www.kegg.jp/kegg/docs/keggapi.html Table 1)
idType <- "ENSEMBL"
returnType <- "GeneSetCollection" # c("GeneSetCollection", "list")
hsa <- getGenesets(org = org, db = db, gene.id.type = idType, cache = TRUE, return.type = returnType)
pwys <- keggList(database = kegg_db, organism = org)
```
```{r test}
print("hsa info")
#print(attributes(hsa[[1]]))
geneSet <- hsa[[1]]
pwyinfo <- pwy[[1]]
#print(attributes(pwyinfo))
#tmp <- hsa[[1]]$geneIds
#print("pwy unlist")
#gene_id <- unlist(hsa, use.names = TRUE)
#gene_id_mod <- unlist(hsa, use.names = FALSE)
#print(unlist(hsa, use.names=FALSE))
# step5: Parse and write the gene sets to a flat text file in GMT format for other pathway enrichment analysis programs (e.g., GSEA):
#writeGMT(hsa, gmt.file = "20220424_kegg_hsa_gmt")
# library("MEGENA")
# output.geneSet.file(unlist(hsa, TRUE, TRUE), "kegg_hsa-2022-04-25.gmt")
print('hello world')
pathway.codes <- sub("path:", "", names(pwys))
pathway.class <- sapply(pathway.codes,
                           function(pwid) {
                             pw <- keggGet(pwid)
                             return(pw[[1]]$CLASS)
                           }
)
pathway.name <- sapply(pathway.codes,
                           function(pwid) {
                             pw <- keggGet(pwid)
                             return(pw[[1]]$NAME)
                           }
)
remove(tmp)
tmp <- keggGet("hsa00010")
#print(kegg.path.info <- attributes(unlist(tmp)))
print(tmp[[1]]$CLASS)
el <- tmp[[1]]
p <- pwy[[1]]
ids <- attr(geneSet, "geneIds")
print(list(pathway.codes[[1]], el$CLASS, el$NAME, paste(ids, collapse = ";"), length(ids)))
```

```{r gene_id}
pathway.codes <- sub("path:", "", names(pathways.list))
genes.by.pathway <- sapply(pathway.codes,
                           function(pwid) {
                             pw <- keggGet(pwid)
                             if (is.null(pw[[1]]$GENE)) return(NA)
                             pw2 <- pw[[1]]$GENE[c(TRUE, FALSE)] # may need to modify this to c(FALSE, TRUE) for other organisms
                             pw2 <- unlist(lapply(strsplit(pw2, split = ";", fixed = T), function(x)x[1]))
                             return(pw2)
                           }
)
head(genes.by.pathway)
```
```{r check and obtain a list of entry identifiers}
# step2: check and obtain a list of entry identifiers (in this case: hsa) and associated definition for a given database or a given set of database entries.
human <- keggList(database = kegg_db, organism = org)

```

```{r get known pathways}
# step 3: download the pathways of that organism
hsapathways <- downloadPathways(org)
```
```{r retrieve KEGG pathways}
# step 4: retrieve gene sets for an organism from databases such as GO and KEGG:
#library("EnrichmentBrowser")
idType <- "ENSEMBL"
returnType <- "list" # c("GeneSetCollection", "list")
hsa <- getGenesets(org = org, db = db, gene.id.type = idType, cache = TRUE, return.type = returnType)
```

