```{r setup & libraries}
# load and install corresponding librarys
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager", INSTALL_opts = '--no-lock')
if (!requireNamespace("MEGENA", quietly = TRUE))
  install.packages("MEGENA")
#if (!requireNamespace("data.table", quietly = TRUE))
#  install.packages('data.table')
BiocManager::install("EnrichmentBrowser", ask = FALSE)
library("EnrichmentBrowser")
if (!isAvailable("KEGGREST"))
  BiocManager::install("KEGGREST", ask = FALSE, type = "source", checkBuilt = TRUE)

library("KEGGREST")
```

```{r query and parse }
# library("EnrichmentBrowser")

db <- "kegg"
org <- "hsa"
kegg_db <- "pathway" # KEGG database anme ie. pathway, brite, module, orthology, genome, genes
# list can be found at (https://www.kegg.jp/kegg/docs/keggapi.html Table 1)
idType <- "ENSEMBL"
returnType <- "GeneSetCollection" # c("GeneSetCollection", "list")
hsa <- getGenesets(org = org, db = db, gene.id.type = idType, cache = TRUE, return.type = returnType)
pwys <- keggList(database = kegg_db, organism = org)

```
```{r create and save dataframe}
remove(PathwayID, PathwayClass, PathwayName)
PathwayID <- sub("path:", "", names(pwys))
pws <- unname(sapply(PathwayID,
                     function(pwid) return(keggGet(pwid))))
PathwayClass <- sapply(pws,
                       function(pw) {
                         if (is.null(pw$CLASS)) return(NA)
                         return(pw$CLASS)
                       })

PathwayName <- sapply(pws,
                      function(pw) {
                        return(pw$NAME)
                      })

All_Genes <- sapply(hsa,
                    function(geneset) {
                      ids <- attr(geneset, "geneIds")
                      if (is.null(ids)) return(NA)
                      return(paste(ids, collapse = ";"))
                    })

NumberOfGenes <- sapply(hsa,
                        function(geneset) {
                          ids <- attr(geneset, "geneIds")
                          return(length(ids))
                        })

#print(c(length(PathwayID), length(PathwayClass), length(PathwayName), length(All_Genes), length(NumberOfGenes)))
#df <- data.frame(PathwayID, unname(PathwayClass), unname(PathwayName), All_Genes, NumberOfGenes)
#head(df)
#write.csv(df,"pathways-27-Apr-2022-1115.csv", row.names = FALSE)
```
```{r playing}
df <- data.frame("PathwayId" = PathwayID, "PathwayClass" = PathwayClass, "PathwayName" = PathwayName, "All_Genes" = All_Genes, "NumberOfGenes" = NumberOfGenes)
head(df)
write.csv(df,"pathways-27-Apr-2022-1115.csv", row.names = FALSE)
test <- tmp[[1]]$CLASS
print(test)
test <- unname(PathwayName)
print(str(unname(PathwayName)))
str(PathwayClass)
str(NumberOfGenes)
#print(PathwayClass[1:347])
#print(PathwayName[1:347])
#print(PathwayName[1:347])
```

```{r gene_id}
pathway.codes <- sub("path:", "", names(pathways.list))
genes.by.pathway <- sapply(pathway.codes,
                           function(pwid) {
                             pw <- keggGet(pwid)
                             if (is.null(pw[[1]]$GENE)) return(NA)
                             pw2 <- pw[[1]]$GENE[c(TRUE, FALSE)] # may need to modify this to c(FALSE, TRUE) for other organisms
                             pw2 <- unlist(lapply(strsplit(pw2, split = ";", fixed = T), function(x)x[1]))
                             return(pw2)
                           }
)
head(genes.by.pathway)
```
```{r check and obtain a list of entry identifiers}
# step2: check and obtain a list of entry identifiers (in this case: hsa) and associated definition for a given database or a given set of database entries.
human <- keggList(database = kegg_db, organism = org)

```

```{r get known pathways}
# step 3: download the pathways of that organism
hsapathways <- downloadPathways(org)
```
```{r retrieve KEGG pathways}
# step 4: retrieve gene sets for an organism from databases such as GO and KEGG:
#library("EnrichmentBrowser")
idType <- "ENSEMBL"
returnType <- "list" # c("GeneSetCollection", "list")
hsa <- getGenesets(org = org, db = db, gene.id.type = idType, cache = TRUE, return.type = returnType)
```

